Add support for tokens requested via oidc-agent.

bbockelm opened the pull request:
Example invocation:

```
./bin/smoke-test.sh -o WLCG-XFER -t wlcg https://red-gridftp1.unl.edu:1094/user/dteam
```

(assuming that you have already run `oidc-add WLCG-XFER`)

bbockelm committed:
Add support for tokens requested via oidc-agent.

bbockelm committed:
Cleanup error handling a bit.

bbockelm committed:
Add ability to request tokens in tests and use alternate endpoint files.

bbockelm committed:
Add some sample endpoints.

bbockelm committed:
Adjust token endpoints.

bbockelm committed:
Allow env var override of VOMS VO

bbockelm committed:
Add manchester endpoint

bbockelm commented:
Ok, I am done hacking for the day!  Ready for review...

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -58,7 +58,7 @@ extended=0
 tokenType=macaroon
```
[Unfortunately, githib doesn't allow me to make this comment any earlier]

I think the patch should define `WLCG_THIRDPARTY_PRIVATE_URL` and `WLCG_THIRDPARTY_UPLOAD_BASE_URL` (as with `dteam`, `escape` and `atlas` VOs), and update the code to use these default values if the client configures smoke-test to use the `wlcg` VO.

bbockelm replied:
Hi @paulmillar -

Finally started to look at this and I noticed that something appears to be misconfigured on the prometheus side.

This works:
```
voms-proxy-init -voms dteam
curl --verbose --connect-timeout 60 -D /tmp/tmp.CNpXBTEwRC \
   -s -f -L  --capath /etc/grid-security/certificates -H 'X-No-Delegate: true' \
   --cacert /tmp/x509up_u1221 -E /tmp/x509up_u1221 -H 'Credential: none' \
   -m30 -X POST -H 'Content-Type: application/macaroon-request' \
   -d '{"caveats": ["activity:DOWNLOAD"], "validity": "PT30M"}' -o/tmp/tmp.FbihlmyRjK \
   https://prometheus.desy.de:2443/VOs/dteam/private-file
```

This doesn't work:
```
voms-proxy-init -voms wlcg
curl --verbose --connect-timeout 60 -D /tmp/tmp.CNpXBTEwRC \
   -s -f -L  --capath /etc/grid-security/certificates -H 'X-No-Delegate: true' \
   --cacert /tmp/x509up_u1221 -E /tmp/x509up_u1221 -H 'Credential: none' \
   -m30 -X POST -H 'Content-Type: application/macaroon-request' \
   -d '{"caveats": ["activity:DOWNLOAD"], "validity": "PT30M"}' -o/tmp/tmp.FbihlmyRjK \
   https://prometheus.desy.de:2443/VOs/wlcg/private-file
```

Only difference is replacing the `dteam` string with `wlcg`.  Access with DAVIX clients appears OK, so I think X.509 auth is working overall -- but perhaps just an issue with the macaroon generation?

paulmillar replied:
Yes, indeed.  The problem was that prometheus was missing the .lsc file for the WLCG VOMS server.

I've installed the missing file, so the problem should be fixed now.

Could you give it another go?

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -58,7 +58,7 @@ extended=0
 workarounds=""
 debugOutput=0
-vo=dteam
+vo=${SMOKE_TEST_VO:-dteam}
```
This is OK, but I'm wondering if it's necessary.

It is (I believe) redundant, as the VO may be specified by the `-v` command-line option.

The `smoke-em.sh` script doesn't seem to set this environment variable, so I'm guessing the intention is that the `SMOKE_TEST_VO` environment variable is set external to both scripts.

An alternative approach would be to add a `-v` option to `smoke-em.sh` script, where it simply forwards the supplied option value to `smoke-test.sh`.

I don't have a strong opinion on which approach is better, but note that currently (all?) other options are set via explicit command-line arguments and that the command-line arguments are documented through the `-h`/`-?` help. ... so perhaps I'm leaning more towards preferring command-line argument.

bbockelm replied:
Yup, agreed..  Will resolve this.

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -528,6 +528,34 @@ requestMacaroon() { # $1 Caveats, $2 URL, $3 variable for macaroon, $4 variable
     eval $3="$macaroon"
 }
 
+requestWlcgToken() { # $1 oidc-agent account name, $2 audience, $3 variable for token, $4 token for result.
```
Very minor comment, but would this be better as `requestOidcAgentToken` (or similar)?

There doesn't seem to be anything specific to WLCG-Profile here.

bbockelm replied:
Sure, good idea.

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -528,6 +528,34 @@ requestMacaroon() { # $1 Caveats, $2 URL, $3 variable for macaroon, $4 variable
+
+    echo -n "Requesting token from account $1 for audience $2: "
+
+    token=$(oidc-token $1 --aud $2)
```
I would suggest putting the aud argument in quotes; e.g.,

    token=$(oidc-token $1 --aud "$2")

This allows `$2` to be empty, which `oidc-agent` interprets as a request to use the OP's default audience claim.

Also, `$1` could be put in double-quote-marks, in case someone has an account shortname that has a space in it.

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -592,13 +620,14 @@ while getopts "h?t:v:p:r:u:s:fxlLCcd2" opt; do
+            echo "  -o <name>  oidc-agent account name (for use with wlcg tokens)"
             echo
             echo "Defaults:"
             echo "  -p $THIRDPARTY_UNAUTHENTICATED_URL"
```
[github doesn't let me add this comment any further down in the file]

As above, if you add support for default WLCG VO values, please also document them here.

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -592,13 +620,14 @@ while getopts "h?t:v:p:r:u:s:fxlLCcd2" opt; do
             echo "  -L         Enable location-trusted work-around"
             echo "  -2         Force client not to use TLS v1.3 as work-around"
-            echo "  -t <token> Use <token> for non-X.509 authn to target"
+            echo "  -t <token> Use <token> for non-X.509 authn to target (valid values: macaroon, scitoken, wlcg)"
```
I'm wondering whether `wlcg` is the best choice of name here.

From what I can see, the added support is more general that this.  Any token returned by oidc-agent is used (which might or might not be an WLCG-AuthZ-WG-Profile).  If there's any requirement, it's that the token must be a JWT.

Perhaps `oidc-agent` instead of `wlcg`?

paulmillar commented on bin/smoke-test.sh:
```diff
@@ -655,6 +687,9 @@ while getopts "h?t:v:p:r:u:s:fxlLCcd2" opt; do
+            ;;
         p)
             THIRDPARTY_UNAUTHENTICATED_URL=$OPTARG
             ;;
```
[github doesn't let me make this comment further down the file]

Line 748 has:

    [ "$tokenType" = "scitoken" ] && [ "$sciTokenServer" = "" ] && fatal "No SciToken server"

There should be a similar test to verify that if `tokenType` is `WlcgToken` then `OIDC_AGENT_ACCOUNT` has also been specified.

bbockelm replied:
Ah - indeed, and the above is actually a bug too.  Should be `[ "$tokenType" = "SciToken" ]`; I'll fix it.

paulmillar reviewed (commented):
I'm happy with the patch as-is.

I've made a few commands.  There are no show-stoppers, but would appreciate it if you could check the comments.

bbockelm committed:
Address comments from code review

- Add new WLCG-specific default URLs.
- Make the VO be set by the `-v` option.
- Rename `wlcg` token type to `oidcagent`.
- Make it an error to set the `oidcagent` token type but not specify an account name.

bbockelm commented:
Ok, I think I got the comments addressed - except for this issue (I think) with the WLCG VO requesting macaroons noted above.
